### PR TITLE
[core] Wire processing stages into SourceCodeProcessor

### PR DESCRIPTION
**Pull request by oowekyala**
Follow up on #1426 (including PR comments from the later [reviews](https://github.com/pmd/pmd/pull/1426#pullrequestreview-191959120))

This is a temporary solution until we change RuleSet. You can see when running with the benchmark that the stages are properly executed for any language (though the benchmark reporter doesn't preserve the order of execution).

With this, plus #1759, we can plug-in a naive name disambiguation stage, which unlocks progress on the updating of type resolution (and on a long-term solution for stages). This is what I've tried to schematize [on the wiki](https://github.com/pmd/pmd/wiki/PMD-7.0.0-Java)

### CODE REVIEW

**Comment by adangel** [comment]
By the way, I figured out why travis doesn't build, even though the PR is build on the merge commit with pmd/7.0.x : the important change is in `.travis.yml`, but the git checkout/merge happens only after travis is already running, so it doesn't change the receipt. Since there are no commit comments on this PR on github yet, I'd suggest to rebase this PR and force push:

```
git fetch origin
git checkout wire-processing-stages
git rebase origin/pmd/7.0.x
git push -f wire-processing-stages
```

You'll get rid of the two "Merge branch '7.0.x' into wire-processing-stages" commits then. If you needed to resolve conflicts back then, you probably need to resolve them again during rebase.

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Running pmdtester failed, this message is mainly used to remind the maintainers of PMD.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by djydewang** [comment]
> Running pmdtester failed, this message is mainly used to remind the maintainers of PMD.

https://travis-ci.org/pmd/pmd/jobs/526201635#L6720, it seems that the server has no local pmd/7.0.x branch. I'll fix the issue in another PR.

File: pmd-vm/src/main/java/net/sourceforge/pmd/lang/vm/ast/ASTprocess.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.vm.ast;
+
+import net.sourceforge.pmd.lang.ast.RootNode;
+
+public final class ASTprocess extends AbstractVmNode implements RootNode {
```
Comment (jsotuyod) [inline pmd-vm/src/main/java/net/sourceforge/pmd/lang/vm/ast/ASTprocess.java]: where does this come from? I don't see the VM grammar being edited…

Comment (oowekyala) [reply pmd-vm/src/main/java/net/sourceforge/pmd/lang/vm/ast/ASTprocess.java]: I made it implement RootNode (needed by AST stages).

File: pmd-core/src/main/java/net/sourceforge/pmd/internal/RulesetStageDependencyHelper.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+/**
+ * Temporary helper for {@link SourceCodeProcessor} until we change rulesets.
+ * See comment in SourceCodeProcessor.
+ *
+ * @author Clément Fournier
+ */
+public class RulesetStageDependencyHelper {
+
+    // cache by ruleset
+    private final Map<RuleSets, Map<LanguageVersion, List<AstProcessingStage<?>>>> dependenciesByRuleset = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/internal/RulesetStageDependencyHelper.java]: This is supposed to be a temporary implementation, eventually `RuleSets` should be removed and `RuleSet` should adopt this logic in some way. So I don't think we should care too much here if they are duplicated entries.

I committed your suggestion about using concurrent maps though.

**Pull request merged**